pkg/git/changelog.go

```diff
@@ -0,0 +1,267 @@
+	BreakingChange bool
+	StoryLink      string
+	IssueLink      string
+}
+
+type GitChangeLog struct {
+	VersionBump string
+	Changes     GitChanges
+	OutputMessage string
+}
+
+type GitChangeLogOptions struct {
+	Description bool
+	UnknownType bool
+}
+
+type GitChanges []GitChange
+
+var minorTypes = map[string]bool{"feat": true, "perf": true, "test": true, "chore": true}
+var patchTypes = map[string]bool{"fix": true, "docs": true, "style": true, "refactor": true}
+
+func (g GitWrapper) ChangeLog(logPath string, command string, svc issues.IssueService, options GitChangeLogOptions) (GitChangeLog, error) {
+	args := append([]string{"-C", g.dir, "log"}, logPath, command)
+	out, err := pkg.NewCommand("git", args...).RunOut()
+	commits := regexp.MustCompile(`commit .{40}`).Split(out, -1)
+	repo := GetRepo(g.dir)
+	owner := GetOwner(g.dir)
+	gitChangesHolder := make([]GitChange, len(commits)-1)
+
+	for index, commit := range commits[1:] {
+		lines := regexp.MustCompile(`\n`).Split(commit, -1)
+		if len(lines) < 5 {
+			continue
+		}
+
+		title := GetCommitTitle(lines)
+		body := GetBody(lines)
+
+		if !CommitIsValid(title) {
+			change := GitChange{
+				Body:           body,
+			}
+			gitChangesHolder[index] = change
+			continue
+		}
+
+		committer := GetCommitter(lines)
+		commitType := GetCommitType(title)
+		breakingChangeFound := CommitHasBreakingChange(body)
+		issueNum := GetIssueNum(body)
+		issueLink := GetIssueLink(g.dir, issueNum)
+		storyLink, _ := GetStoryLink(svc, owner, repo, issueNum)
+
+		change := GitChange{
+			Valid:          true,
+			Title:          title,
+			Body:           body,
+			Issue:          issueNum,
+			IssueLink: 		issueLink,
+			StoryLink:      storyLink,
+			Committer:      committer,
+			CommitType:     commitType,
+			BreakingChange: breakingChangeFound,
+		}
+
+		gitChangesHolder[index] = change
+
+	}
+
+	changes := GitChanges(gitChangesHolder)
+	versionBump := changes.GetVersionBump()
+	changeLog := GitChangeLog{
+		VersionBump: versionBump,
+		Changes:     changes,
+		OutputMessage: GetChangeLogOutputMessage(versionBump, changes, options),
+	}
+
+	return changeLog, err
+}
+
+func CommitIsValid(title string) bool {
+	valid, _ := regexp.MatchString(`.*(\()(.)*(\))(:){1}`, title)
+	return valid
+}
+
+func CommitHasBreakingChange(body string) bool {
+	var breakingChangeFound = true
+	breakingChangeFormat := regexp.MustCompile(`BREAKING CHANGE: .*\n`)
+	breakingChange := breakingChangeFormat.FindAllString(body, -1)
+
+	if len(breakingChange) == 0 {
+		breakingChangeFound = false
+	}
+	return breakingChangeFound
+}
+
+func (g GitChanges) GetVersionBump() string {
+	majorChanges, minorChanges, patchChanges, _ := g.GetSeparatedChanges()
+
+	if len(majorChanges) > 0 {
+		return "MAJOR"
+	} else if len(minorChanges) > 0 {
+		return "MINOR"
+	} else if len(patchChanges) > 0 {
+		return "PATCHES"
+	}
+	return "NONE"
+}
+
+func (g GitChanges) GetSeparatedChanges() ([]GitChange, []GitChange, []GitChange, []GitChange) {
```

Comment:
Another way to do this would be to define constants for the bump levels (`const Major = "major"` etc), then put the changes into a `map[string][]GitChanges`. Adding things to a slice in a map is easy in go: `m[x] = append(m[x],v)` works even if you've never set `m[x]`, because you can append to a nil slice.